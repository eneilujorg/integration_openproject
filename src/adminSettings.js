/* jshint esversion: 6 */

/**
 * SPDX-FileCopyrightText: 2021-2023 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import Vue from 'vue'
import './bootstrap.js'
import AdminSettings from './components/AdminSettings.vue'

// eslint-disable-next-line
'use strict'

// eslint-disable-next-line
new Vue({
	el: '#openproject_prefs',
	render: h => h(AdminSettings),
})
