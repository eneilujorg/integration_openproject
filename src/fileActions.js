/**
 * SPDX-FileCopyrightText: 2023 Nextcloud GmbH and Nextcloud contributors
 * SPDX-FileCopyrightText: 2022 Jankari Tech Pvt. Ltd.
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import './bootstrap.js'

OCA.Files.fileActions.registerAction({
	name: 'open-project',
	displayName: t('integration_openproject', 'OpenProject'),
	mime: 'all',
	permissions: OC.PERMISSION_READ,
	iconClass: 'icon-openproject',
	actionHandler: (filename, context) => {
		const fileList = context.fileList
		if (!fileList._detailsView) {
			return
		}
		// use the sidebar-tab id for the navigation
		fileList.showDetailsView(filename, 'open-project')
	},
})
