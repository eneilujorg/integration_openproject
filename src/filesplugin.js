/*
 * Copyright (c) 2023 Sagar Gurung <sagar@jankaritech.com>
 *
 * This file is licensed under the Affero General Public License version 3
 * or later.
 *
 * See the COPYING-README file.
 *
 */
import './bootstrap.js'
import Vue from 'vue'
import LinkMultipleFilesModal from './views/LinkMultipleFilesModal.vue'

(function() {
	if (!OCA.OpenProject) {
		/**
		 * @namespace
		 */
		OCA.OpenProject = {
			requestOnFileChange: false,
		}
	}

	/**
	 * @namespace
	 */
	OCA.OpenProject.FilesPlugin = {
		ignoreLists: [
			'trashbin',
			'files.public',
		],

		attach(fileList) {
			if (this.ignoreLists.indexOf(fileList.id) >= 0) {
				return
			}
			fileList.registerMultiSelectFileAction({
				name: 'open-project',
				displayName: t('integration_openproject', 'Link to work package'),
				mime: 'all',
				permissions: OC.PERMISSION_READ,
				iconClass: 'icon-openproject',
				action: (selectedFiles) => { this.signExample(selectedFiles) },
			})
		},
		signExample: (selectedFiles) => {
			// store all the file-id in an array and set the file ids
			const fileInfos = []
			for (const file of selectedFiles) {
				const fileInfo = {
					id: file.id,
					name: file.name,
				}
				fileInfos.push(fileInfo)
			}
			OCA.OpenProject.LinkMultipleFilesModalVue.$children[0].setFileInfos(fileInfos)
			OCA.OpenProject.LinkMultipleFilesModalVue.$children[0].showModal()
		},
	}
})()

OC.Plugins.register('OCA.Files.FileList', OCA.OpenProject.FilesPlugin)

const modalId = 'exampleModal'
const modalElement = document.createElement('div')
modalElement.id = modalId
document.body.append(modalElement)

OCA.OpenProject.LinkMultipleFilesModalVue = new Vue({
	el: modalElement,
	render: h => {
		return h(LinkMultipleFilesModal)
	},
})
