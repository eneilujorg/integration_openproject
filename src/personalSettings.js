/* jshint esversion: 6 */

/**
 * SPDX-FileCopyrightText: 2021-2023 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import Vue from 'vue'
import './bootstrap.js'
import PersonalSettings from './components/PersonalSettings.vue'

// eslint-disable-next-line
'use strict'

// eslint-disable-next-line
new Vue({
	el: '#openproject_prefs',
	render: h => h(PersonalSettings),
})
