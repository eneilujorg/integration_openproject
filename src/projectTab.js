/**
 * SPDX-FileCopyrightText: 2023 Nextcloud GmbH and Nextcloud contributors
 * SPDX-FileCopyrightText: 2021-2022 Jankari Tech Pvt. Ltd.
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import Vue from 'vue'

import './bootstrap.js'
import ProjectsTab from './views/ProjectsTab.vue'

// Init OpenProject Tab Service
if (!window.OCA.OpenProject) {
	window.OCA.OpenProject = {}
}

const View = Vue.extend(ProjectsTab)
let TabInstance = null

const projectTab = new OCA.Files.Sidebar.Tab({
	id: 'open-project',
	name: t('integration_openproject', 'OpenProject'),
	icon: 'icon-openproject',

	async mount(el, fileInfo, context) {
		if (TabInstance) {
			TabInstance.$destroy()
		}
		TabInstance = new View({
			// Better integration with vue parent component
			parent: context,
		})
		// Only mount after we have all the info we need
		await TabInstance.update(fileInfo)
		TabInstance.$mount(el)
	},
	update(fileInfo) {
		TabInstance.update(fileInfo)
	},
	destroy() {
		TabInstance.$destroy()
		TabInstance = null
	},
})

window.addEventListener('DOMContentLoaded', function() {
	if (OCA.Files && OCA.Files.Sidebar) {
		OCA.Files.Sidebar.registerTab(projectTab)
	}
})
