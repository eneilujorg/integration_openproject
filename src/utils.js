let mytimer = 0
export function delay(callback, ms) {
	return function() {
		const context = this
		const args = arguments
		clearTimeout(mytimer)
		mytimer = setTimeout(function() {
			callback.apply(context, args)
		}, ms || 0)
	}
}

export const STATE = {
	OK: 'ok',
	ERROR: 'error',
	LOADING: 'loading',
	NO_TOKEN: 'no-token',
	CONNECTION_ERROR: 'connection-error',
	FAILED_FETCHING_WORKPACKAGES: 'failed-fetching-workpackages',
}
