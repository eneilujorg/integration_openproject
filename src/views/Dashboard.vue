<template>
	<DashboardWidget :items="items"
		:show-more-url="showMoreUrl"
		:show-more-text="title"
		:loading="isLoading">
		<template #empty-content>
			<EmptyContent v-if="emptyContentMessage"
				:icon="emptyContentIcon">
				<template #desc>
					<div v-if="!!requestUrl">
						{{ emptyContentMessage }}
					</div>
					<div v-if="showOauthConnect" class="connect-button">
						<OAuthConnectButton :request-url="requestUrl" />
					</div>
				</template>
			</EmptyContent>
		</template>
	</DashboardWidget>
</template>

<script>
import axios from '@nextcloud/axios'
import { generateUrl } from '@nextcloud/router'
import { DashboardWidget } from '@nextcloud/vue-dashboard'
import { showError } from '@nextcloud/dialogs'
import '@nextcloud/dialogs/styles/toast.scss'
import moment from '@nextcloud/moment'
import EmptyContent from '@nextcloud/vue/dist/Components/EmptyContent'
import { loadState } from '@nextcloud/initial-state'
import OAuthConnectButton from '../components/OAuthConnectButton'
import { STATE } from '../utils'

export default {
	name: 'Dashboard',

	components: {
		DashboardWidget, EmptyContent, OAuthConnectButton,
	},

	props: {
		title: {
			type: String,
			required: true,
		},
	},

	data() {
		return {
			openprojectUrl: null,
			notifications: [],
			loop: null,
			state: STATE.LOADING,
			requestUrl: loadState('integration_openproject', 'request-url'),
			settingsUrl: generateUrl('/settings/user/connected-accounts'),
			themingColor: OCA.Theming ? OCA.Theming.color.replace('#', '') : '0082C9',
			windowVisibility: true,
		}
	},

	computed: {
		isLoading() {
			return this.state === STATE.LOADING
		},
		showMoreUrl() {
			return this.openprojectUrl + '/notifications'
		},
		items() {
			return this.notifications.map((n) => {
				return {
					id: this.getUniqueKey(n),
					targetUrl: this.getNotificationTarget(n),
					avatarUrl: this.getAuthorAvatarUrl(n),
					avatarUsername: this.getAuthorShortName(n) + 'z',
					overlayIconUrl: this.getNotificationTypeImage(n),
					mainText: this.getTargetTitle(n),
					subText: this.getSubline(n),
				}
			})
		},
		lastDate() {
			const nbNotif = this.notifications.length
			return (nbNotif > 0) ? this.notifications[0].updatedAt : null
		},
		lastMoment() {
			return moment(this.lastDate)
		},
		emptyContentMessage() {
			if (this.state === STATE.NO_TOKEN) {
				return t('integration_openproject', 'No connection with OpenProject')
			} else if (this.state === STATE.ERROR) {
				return t('integration_openproject', 'Error connecting to OpenProject')
			} else if (this.state === STATE.OK) {
				return t('integration_openproject', 'No OpenProject notifications!')
			}
			return ''
		},
		emptyContentIcon() {
			if (!this.requestUrl) {
				return 'icon-noConnection'
			} else if (this.state === STATE.CONNECTION_ERROR) {
				return 'icon-noConnection'
			} else if (this.state === STATE.ERROR) {
				return 'icon-close'
			} else if (this.state === STATE.OK) {
				return 'icon-checkmark'
			}
			return 'icon-checkmark'
		},
		showOauthConnect() {
			return [STATE.NO_TOKEN, STATE.ERROR].includes(this.state)
		},
	},

	watch: {
		windowVisibility(newValue) {
			if (newValue) {
				this.launchLoop()
			} else {
				this.stopLoop()
			}
		},
	},

	beforeDestroy() {
		document.removeEventListener('visibilitychange', this.changeWindowVisibility)
	},

	beforeMount() {
		this.launchLoop()
		document.addEventListener('visibilitychange', this.changeWindowVisibility)
	},

	methods: {
		changeWindowVisibility() {
			this.windowVisibility = !document.hidden
		},
		stopLoop() {
			clearInterval(this.loop)
		},
		async launchLoop() {
			if (!this.requestUrl) {
				this.state = STATE.ERROR
				return
			}
			// get openproject URL first
			try {
				const response = await axios.get(generateUrl('/apps/integration_openproject/url'))
				this.openprojectUrl = response.data.replace(/\/+$/, '')
			} catch (error) {
				console.debug(error)
			}
			// then launch the loop
			this.fetchNotifications()
			this.loop = setInterval(() => this.fetchNotifications(), 60000)
		},
		fetchNotifications() {
			const req = {}
			if (this.lastDate) {
				req.params = {
					since: this.lastDate,
				}
			}
			axios.get(generateUrl('/apps/integration_openproject/notifications'), req).then((response) => {
				if (Array.isArray(response.data)) {
					this.processNotifications(response.data)
				}
				this.state = STATE.OK
			}).catch((error) => {
				clearInterval(this.loop)
				if (error.response && error.response.status === 400) {
					this.state = STATE.NO_TOKEN
				} else if (error.response && error.response.status === 401) {
					showError(t('integration_openproject', 'Failed to get OpenProject notifications'))
					this.state = STATE.ERROR
				} else {
					// there was an error in notif processing
					console.debug(error)
				}
			})
		},
		processNotifications(newNotifications) {
			if (this.lastDate) {
				// just add those which are more recent than our most recent one
				let i = 0
				while (i < newNotifications.length && this.lastMoment.isBefore(newNotifications[i].updatedAt)) {
					i++
				}
				if (i > 0) {
					const toAdd = this.filter(newNotifications.slice(0, i))
					this.notifications = toAdd.concat(this.notifications)
				}
			} else {
				// first time we don't check the date
				this.notifications = this.filter(newNotifications)
			}
		},
		filter(notifications) {
			return notifications
		},
		getNotificationTarget(n) {
			const wpId = n._links?.resource?.href
				? n._links.resource.href.replace(/.*\//, '')
				: null
			return wpId
				? this.openprojectUrl + '/notifications/details/' + wpId + '/activity/'
				: ''
		},
		getUniqueKey(n) {
			return n.id + ':' + n.updatedAt
		},
		getAuthorShortName(n) {
			return n._links?.actor?.title
				? n._links.actor.title
				: undefined
		},
		getAuthorFullName(n) {
			return n.firstname + ' ' + n.lastname
		},
		getAuthorAvatarUrl(n) {
			const userId = n._links?.actor?.href
				? n._links.actor.href.replace(/.*\//, '')
				: null
			const userName = n._links?.actor?.title
				? n._links.actor.title
				: null
			return userId
				? generateUrl('/apps/integration_openproject/avatar?') + encodeURIComponent('userId') + '=' + userId + '&' + encodeURIComponent('userName') + '=' + userName
				: ''
		},
		getNotificationProjectName(n) {
			return ''
		},
		getNotificationContent(n) {
			return ''
		},
		getNotificationTypeImage(n) {
			return generateUrl('/svg/core/actions/sound?color=' + this.themingColor)
		},
		getSubline(n) {
			return n._links.project.title + ' - ' + t('integration_openproject', n.reason)
		},
		getTargetTitle(n) {
			return n._links.resource.title
		},
		getFormattedDate(n) {
			return moment(n.updated_at).format('LLL')
		},
	},
}
</script>

<style scoped lang="scss">
::v-deep .connect-button {
	margin-top: 10px;
}
</style>
