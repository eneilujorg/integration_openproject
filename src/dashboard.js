/**
 * SPDX-FileCopyrightText: 2021-2023 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import Vue from 'vue'
import './bootstrap.js'
import Dashboard from './views/Dashboard.vue'

document.addEventListener('DOMContentLoaded', function() {

	OCA.Dashboard.register('openproject_notifications', (el, { widget }) => {
		const View = Vue.extend(Dashboard)
		new View({
			propsData: { title: widget.title },
		}).$mount(el)
	})

})
