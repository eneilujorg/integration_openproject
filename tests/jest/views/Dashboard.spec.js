/* jshint esversion: 8 */

import { shallowMount, mount, createLocalVue } from '@vue/test-utils'
import Dashboard from '../../../src/views/Dashboard'
import axios from '@nextcloud/axios'
import * as initialState from '@nextcloud/initial-state'
import { STATE } from '../../../src/utils'
import notificationsResponse from '../fixtures/notificationsResponse.json'
import * as dialogs from '@nextcloud/dialogs'

jest.mock('@nextcloud/axios')
jest.mock('@nextcloud/l10n', () => ({
	translate: jest.fn((app, msg) => msg),
	getLanguage: jest.fn(),
}))
jest.mock('@nextcloud/dialogs', () => ({
	getLanguage: jest.fn(() => ''),
	showError: jest.fn(),
	showSuccess: jest.fn(),
}))

global.OCA = {}
global.OC = {}
const localVue = createLocalVue()

describe('Dashboard.vue', () => {
	const dashboardTriggerButtonSelector = '.trigger'
	const markAsReadButtonSelector = '.popover__wrapper .action-button'
	let wrapper
	beforeEach(() => {
		// eslint-disable-next-line no-import-assign
		initialState.loadState = jest.fn(() => true)

		// mock the beforeMount() method, so that the loop is not called automatically
		// we first need to mount the component, and the tests will call the loop themselves
		Dashboard.beforeMount = jest.fn()
	})
	it('should show the notification items in the Dashboard', async () => {
		wrapper = shallowMount(
			Dashboard,
			{
				localVue,
				mocks: {
					t: (app, msg) => msg,
					generateUrl() {
						return '/'
					},
				},
				propsData: {
					title: 'dashboard',
				},
			})
		const axiosSpy = jest.spyOn(axios, 'get')
			.mockImplementationOnce(() => Promise.resolve({ data: 'http://openproject.org' }))
			.mockImplementationOnce(() => Promise.resolve({ data: notificationsResponse }))
		await wrapper.vm.launchLoop()
		expect(axiosSpy).toBeCalledWith(
			'http://localhost/apps/integration_openproject/url',
		)
		expect(axiosSpy).toBeCalledWith(
			'http://localhost/apps/integration_openproject/notifications',
		)
		expect(wrapper.vm.state).toBe(STATE.OK)
		expect(wrapper.vm.items).toMatchSnapshot()
		axiosSpy.mockRestore()
	})
	describe('mark as read', () => {
		let axiosSpyGet
		beforeEach(() => {
			wrapper = mount(
				Dashboard,
				{
					localVue,
					attachTo: document.body,
					mocks: {
						t: (app, msg) => msg,
						generateUrl() {
							return '/'
						},
					},
					stubs: {
						Avatar: true,
					},
					propsData: {
						title: 'dashboard',
					},
				})
			axiosSpyGet = jest.spyOn(axios, 'get')
				.mockImplementation(() => Promise.resolve({
					data: [
						notificationsResponse[0],
					],
				}))
		})
		afterEach(() => {
			axiosSpyGet.mockRestore()
		})
		it('should mark notifications as read', async () => {
			const axiosSpyDelete = jest.spyOn(axios, 'delete')
				.mockImplementationOnce(() => Promise.resolve({})
				)
			dialogs.showSuccess.mockImplementationOnce()
			await wrapper.vm.fetchNotifications()
			await localVue.nextTick()
			await wrapper.find(dashboardTriggerButtonSelector).trigger('click')
			await localVue.nextTick()
			await wrapper.find(markAsReadButtonSelector).trigger('click')
			await localVue.nextTick()
			expect(axiosSpyDelete).toHaveBeenCalledWith(
				'http://localhost/apps/integration_openproject/work-packages/36/notifications'
			)
			expect(dialogs.showSuccess).toHaveBeenCalledWith(
				'Notifications associated with Work package marked as read'
			)
			wrapper.destroy()
			axiosSpyDelete.mockRestore()
		})
		it('should show an error message if marking as read failed', async () => {
			const axiosSpyDelete = jest.spyOn(axios, 'delete')
				.mockRejectedValueOnce()
			dialogs.showError.mockImplementationOnce()
			await wrapper.vm.fetchNotifications()
			await localVue.nextTick()
			await wrapper.find(dashboardTriggerButtonSelector).trigger('click')
			await localVue.nextTick()
			await wrapper.find(markAsReadButtonSelector).trigger('click')
			await localVue.nextTick()
			expect(dialogs.showError).toHaveBeenCalledWith(
				'Failed to mark notifications as read'
			)
			wrapper.destroy()
			axiosSpyDelete.mockRestore()
		})
	})
})
