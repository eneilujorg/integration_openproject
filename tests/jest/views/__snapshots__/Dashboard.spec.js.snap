// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard.vue should show the notification items in the Dashboard 1`] = `
Array [
  Object {
    "avatarUrl": "http://localhost/apps/integration_openproject/avatar?userId=8&userName=Admin de DEV user",
    "avatarUsername": "Admin de DEV userz",
    "id": "17",
    "mainText": "(5) Create wireframes for new landing page",
    "overlayIconUrl": "",
    "subText": "Scrum project - assigned,mentioned",
    "targetUrl": "http://openproject.org/notifications/details/17/activity/",
  },
  Object {
    "avatarUrl": "http://localhost/apps/integration_openproject/avatar?userId=9&userName=Artur Neumann",
    "avatarUsername": "Artur Neumannz",
    "id": "18",
    "mainText": "(1) Contact form",
    "overlayIconUrl": "",
    "subText": "Scrum project - mentioned",
    "targetUrl": "http://openproject.org/notifications/details/18/activity/",
  },
  Object {
    "avatarUrl": "http://localhost/apps/integration_openproject/avatar?userId=8&userName=Admin de DEV user",
    "avatarUsername": "Admin de DEV userz",
    "id": "36",
    "mainText": "(2) write a software",
    "overlayIconUrl": "",
    "subText": "Dev-large - assigned",
    "targetUrl": "http://openproject.org/notifications/details/36/activity/",
  },
]
`;
